src/zenml/zen_stores/sql_zen_store.py

```diff
@@ -11449,119 +11557,327 @@ def update_tag(
+
+            # If the tag is an exclusive tag, apply the check and attach/detach accordingly
+            if tag_schema.exclusive:
+                scope_ids: Dict[
+                    TaggableResourceTypes, List[Union[UUID, int]]
+                ] = defaultdict(list)
+                detach_resources: List[TagResourceRequest] = []
+
+                if isinstance(resource, PipelineRunSchema):
+                    if resource.pipeline_id:
+                        scope_ids[TaggableResourceTypes.PIPELINE_RUN].append(
+                            resource.pipeline_id
+                        )
+                    other_runs_with_same_tag = self.list_runs(
+                        PipelineRunFilter(
+                            id=f"notequals:{resource.id}",
+                            pipeline_id=resource.pipeline_id,
+                            tags=[tag_schema.name],
+                        )
+                    )
+                    if other_runs_with_same_tag.items:
+                        detach_resources.append(
+                            TagResourceRequest(
+                                tag_id=tag_schema.id,
+                                resource_id=other_runs_with_same_tag.items[
+                                    0
+                                ].id,
+                                resource_type=TaggableResourceTypes.PIPELINE_RUN,
+                            )
+                        )
+                elif isinstance(resource, ArtifactVersionSchema):
+                    if resource.artifact_id:
+                        scope_ids[
+                            TaggableResourceTypes.ARTIFACT_VERSION
+                        ].append(resource.artifact_id)
+                    other_versions_with_same_tag = self.list_artifact_versions(
+                        ArtifactVersionFilter(
+                            id=f"notequals:{resource.id}",
+                            artifact_id=resource.artifact_id,
+                            tags=[tag_schema.name],
+                        )
+                    )
+                    if other_versions_with_same_tag.items:
+                        detach_resources.append(
+                            TagResourceRequest(
+                                tag_id=tag_schema.id,
+                                resource_id=other_versions_with_same_tag.items[
+                                    0
+                                ].id,
+                                resource_type=TaggableResourceTypes.ARTIFACT_VERSION,
+                            )
+                        )
+                elif isinstance(resource, RunTemplateSchema):
+                    scope_ids[TaggableResourceTypes.RUN_TEMPLATE].append(0)
+                    older_templates = self.list_run_templates(
+                        RunTemplateFilter(
+                            id=f"notequals:{resource.id}",
+                            tags=[tag_schema.name],
```

Comment:
This should also filter by pipeline ID I assume?

Reply:
I introduced the proper scoping by pipeline ID as you suggested.